diff --git a/csrc/cumem_allocator.cpp b/csrc/cumem_allocator.cpp
index 78dc840a9..1788c9ea1 100644
--- a/csrc/cumem_allocator.cpp
+++ b/csrc/cumem_allocator.cpp
@@ -107,6 +107,14 @@ void create_and_map(unsigned long long device, ssize_t size, CUdeviceptr d_mem,
   prop.location.id = device;
   prop.allocFlags.compressionType = CU_MEM_ALLOCATION_COMP_NONE;
 
+  int flag = 0;
+  CUDA_CHECK(cuDeviceGetAttribute(
+      &flag, CU_DEVICE_ATTRIBUTE_GPU_DIRECT_RDMA_WITH_CUDA_VMM_SUPPORTED,
+      device));
+  if (flag) {  // support GPUDirect RDMA if possible
+    prop.allocFlags.gpuDirectRDMACapable = 1;
+  }
+
 #ifndef USE_ROCM
   // Allocate memory using cuMemCreate
   CUDA_CHECK(cuMemCreate(p_memHandle, size, &prop, 0));
diff --git a/vllm/model_executor/models/qwen2.py b/vllm/model_executor/models/qwen2.py
index f5501bae7..c6f8c56d3 100644
--- a/vllm/model_executor/models/qwen2.py
+++ b/vllm/model_executor/models/qwen2.py
@@ -174,6 +174,13 @@ class Qwen2Attention(nn.Module):
             if attn_type == AttentionType.ENCODER_ONLY
             else Attention
         )
+        
+        self.disable_attn = False
+        import os
+        if "VLLM_DISABLE_ATTN" in os.environ and os.environ["VLLM_DISABLE_ATTN"] == "1":
+            print("Hijacking the attention in Qwen2 to forbid the use of the attention.")
+            self.disable_attn = True
+        
         self.attn = attn_cls(
             self.num_heads,
             self.head_dim,
@@ -199,7 +206,11 @@ class Qwen2Attention(nn.Module):
         qkv, _ = self.qkv_proj(hidden_states)
         q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
         q, k = self.rotary_emb(positions, q, k)
-        attn_output = self.attn(q, k, v)
+        # For profiling, we hijack here to forbid the use of the attention.
+        if self.disable_attn:
+            attn_output = q
+        else:
+            attn_output = self.attn(q, k, v)
         output, _ = self.o_proj(attn_output)
         return output
 
diff --git a/vllm/v1/engine/async_llm.py b/vllm/v1/engine/async_llm.py
index ec5d6e95c..a507241cc 100644
--- a/vllm/v1/engine/async_llm.py
+++ b/vllm/v1/engine/async_llm.py
@@ -548,6 +548,16 @@ class AsyncLLM(EngineClient):
 
         self.output_handler = asyncio.create_task(output_handler())
 
+    async def abort_all(self) -> int:
+        """Abort all requests in OutputProcessor and EngineCore."""
+        all_request_ids = self.output_processor.abort_all()
+        await self.engine_core.abort_requests_async(all_request_ids)
+
+        if self.log_requests:
+            logger.info("Aborted all requests %s.", ",".join(all_request_ids))
+
+        return len(all_request_ids)
+
     async def abort(self, request_id: str | Iterable[str]) -> None:
         """Abort RequestId in OutputProcessor and EngineCore."""
 
diff --git a/vllm/v1/engine/output_processor.py b/vllm/v1/engine/output_processor.py
index e85fbb4ee..9d7bf9996 100644
--- a/vllm/v1/engine/output_processor.py
+++ b/vllm/v1/engine/output_processor.py
@@ -374,6 +374,11 @@ class OutputProcessor:
             assert state.queue is not None
             state.queue.put(e)
 
+    def abort_all(self) -> list[str]:
+        abort_request_ids = list(self.request_states.keys())
+        self.abort_requests(abort_request_ids)
+        return abort_request_ids
+
     def abort_requests(
         self,
         request_ids: Iterable[str],
diff --git a/vllm/v1/metrics/loggers.py b/vllm/v1/metrics/loggers.py
index dec0e2d00..c3e864a6e 100644
--- a/vllm/v1/metrics/loggers.py
+++ b/vllm/v1/metrics/loggers.py
@@ -1227,11 +1227,14 @@ class StatLoggerManager:
                     custom_prometheus_logger = True
             else:
                 # per engine logger
-                global_stat_logger = PerEngineStatLoggerAdapter(
-                    vllm_config=vllm_config,
-                    engine_indexes=self.engine_indexes,
-                    per_engine_stat_logger_factory=stat_logger_factory,  # type: ignore[arg-type]
-                )
+                if not isinstance(stat_logger_factory, Callable):
+                    global_stat_logger = stat_logger_factory
+                else:
+                    global_stat_logger = PerEngineStatLoggerAdapter(
+                        vllm_config=vllm_config,
+                        engine_indexes=self.engine_indexes,
+                        per_engine_stat_logger_factory=stat_logger_factory,  # type: ignore[arg-type]
+                    )
             self.stat_loggers.append(global_stat_logger)
         if not custom_prometheus_logger:
             self.stat_loggers.append(
diff --git a/vllm/v1/metrics/stats.py b/vllm/v1/metrics/stats.py
index a3078eaa7..ff61c7b12 100644
--- a/vllm/v1/metrics/stats.py
+++ b/vllm/v1/metrics/stats.py
@@ -162,6 +162,10 @@ class KVCacheEvictionEvent:
 @dataclass
 class SchedulerStats:
     """Stats associated with the scheduler."""
+    # NOTE(lhy): we need to add this field to the scheduler stats for PSRL.
+    need_to_abort_reqs: list[str] = field(default_factory=list)
+    req_id_to_prompt_token_num: dict[str, int] | None = None
+    req_id_to_response_token_num: dict[str, int] | None = None
 
     num_running_reqs: int = 0
     num_waiting_reqs: int = 0
